host/s3.go

```diff
@@ -0,0 +1,412 @@
+}
+
+var _DataStream = (*S3DataStream)(nil)
+
+func NewS3DataStream(baseCfg S3DSConfig) (*S3DataStream, error) {
+	cfg, err := baseCfg.ToAWSconfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// create the s3 client
+	// rewrite the default endpoint if it's give (local dev)
+	var s3client *s3.Client
+	if len(baseCfg.Endpoint) > 0 {
+		slog.Warn("local/dev s3 instance")
+		s3client = s3.NewFromConfig(*cfg, func(o *s3.Options) {
+			o.UsePathStyle = true
+			o.BaseEndpoint = aws.String(baseCfg.Endpoint)
+		})
+	} else {
+		s3client = s3.NewFromConfig(*cfg)
+	}
+	// create the trace batcher
+	batcher, err := newTraceBatcher(baseCfg.ByteLimit)
+	if err != nil {
+		return nil, err
+	}
+	return &S3DataStream{
+		config:  baseCfg,
+		client:  s3client,
+		batcher: batcher,
+	}, nil
+}
+
+func (s3ds *S3DataStream) Type() DataStreamType {
+	return DataStreamTypeS3
+}
+
+func (s3ds *S3DataStream) OutputType() DataStreamOutputType {
+	return DataStreamOutputTypeKinesis
+}
+
+func (s3ds *S3DataStream) Start(ctx context.Context) error {
+	slog.Info(
+		"spawning s3 data-stream",
+		slog.Attr{Key: "endpoint", Value: slog.StringValue(s3ds.config.Endpoint)},
+		slog.Attr{Key: "bucket", Value: slog.StringValue(s3ds.config.Bucket)},
+	)
+	opCtx, cancel := context.WithTimeout(ctx, S3ConnectionTimeout)
+	defer cancel()
+
+	err := s3ds.testConnection(opCtx)
+	if err != nil {
+		return err
+	}
+
+	s3ds.spawnPeriodicFlusher(ctx, s3ds.config.FlushInterval)
+
+	mainCtx, mainCancel := context.WithCancel(ctx)
+	s3ds.ctx = mainCtx
+	s3ds.cancelFn = mainCancel
+
+	<-mainCtx.Done()
+	return mainCtx.Err()
+}
+
+func (s3ds *S3DataStream) Stop(ctx context.Context) error {
+	// stop the mainCtx, as we don't want new traces coming in
+	s3ds.cancelFn()
+	// ensure to prune existing traces before closing
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+	s3ds.submitRecords(opCtx)
+	return nil
+}
+
+func (s3ds *S3DataStream) PutRecord(ctx context.Context, event *TraceEvent) error {
+	s3ds.batcher.addNewTrace(event)
+	if s3ds.batcher.isFull() {
+		s3ds.submitRecords(ctx)
+	}
+	return nil
+}
+
+// submitRecords is the private method that will:
+// 1. get the copy of the existing traces, leaving a new batcher empty
+// 2. create the necessary s3 key to submite the batched traces
+// 3. format the traces into a parquet file
+// 4. push the file into the s3 bucket
+// *NOTE*: the method leaves a routine in the background to format the parquet
+//
+//	file, and to submit it into S3. Reason -> to avoid freezing other processes
+func (s3ds *S3DataStream) submitRecords(ctx context.Context) {
+	// check if the trace-buffer has anything
+	if s3ds.batcher.len() <= 0 {
+		return
+	}
+
+	// get and reset the traces from the batcher
+	// increase the counter for the next file descriptor
+	// and get the peer_id from the traces
+	traces := s3ds.batcher.reset()
+	currentFileCnt := s3ds.fileCnt.Load()
+	s3ds.fileCnt.Add(1)
+	producerID := traces[0].PeerID
```

Comment:
There's a tiny race condition here. Let's say `bather.len()` returns 0. Then `batchSize` events are added to the batcher and another go routine calls `submitRecords`. Then the amount of traces could be 0 in the batcher again. So when you call `traces := s3ds.batcher.reset()` you could receive an empty trace list which means `traces[0]` could panic.

This is arguably very unlikely but I'd recommend just getting the traces with `traces := s3ds.batcher.reset()` and check the length before you're doing the slice access.

Reply:
address on latest commits